### PR TITLE
Avoid re-rendering map with "sources"

### PR DESCRIPTION
**Pull request by amitfin**
## Proposed change

[Map card](https://www.home-assistant.io/dashboards/map/) can be configured with [`geo_location_sources`](https://www.home-assistant.io/dashboards/map/#geo_location_sources). It was impossible to interact with the map when [`auto_fit`](https://www.home-assistant.io/dashboards/map/#auto_fit) was also used. The map would reset its view every second (or so).
The fix is to keep the object of the map's entity list, unless there is a real change.

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml
type: map
entities: []
geo_location_sources:
  - geo_json_events
auto_fit: true
```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by karwosts** [comment]
Maybe I'm misunderstanding but this change doesn't look right to me. 

You've replaced updating on hasConfigOrEntitiesChanged with just updating when an entity state string changes (e.g. "home" -> "away"). But this seems like it will reject update when an entity moves around without changing zone, 

The map should redraw anytime one of the watched entities moves, right?

The was another issue I tried to work on once that would have had the map stop autofitting once a user manually interacted with the map, like manually trying to move it. I couldn't quite figure out how to make that work right so I gave up on it. Maybe that would be a better solution for you if it could be made to work?

When you were debugging did you see what was triggering it to rerender constantly, were the geo sources moving or something?

**Comment by amitfin** [comment]
> You've replaced updating on hasConfigOrEntitiesChanged with just updating when an entity state string changes (e.g. "home" -> "away"). But this seems like it will reject update when an entity moves around without changing zone,

You're right. The comparison should be identical to the one that is [here](https://github.com/home-assistant/frontend/blob/a5e5efa5263de790f4624a581f71d6db35eefa65/src/panels/lovelace/common/has-changed.ts#L45) and include all of the state object, including the "latitude" and "longitude" attributes. The code was changed accordingly. Thanks!

> The was another issue I tried to work on once that would have had the map stop auto-fitting once a user manually interacted with the map, like manually trying to move it. I couldn't quite figure out how to make that work right so I gave up on it. Maybe that would be a better solution for you if it could be made to work?

The use case I'm trying to address is geo_location entities which are coming and going (they have a lifetime of 10 minutes). These entities have constant coordinates. We do want to auto-fit the map when there is a new entity, but allow the user to interact with the map (e.g. zoom out on in) while the data on the map hasn't changed.

> When you were debugging did you see what was triggering it to rerender constantly, were the geo sources moving or something?

The debugging was focused on the use case described above ("source" entities with constant coordinates.) However, this change should be correct for any scenario. There is no need to re-render the element when there is no change in the data (causing the element to auto-fit for no reason, when this option is used.)

**Comment by bramkragten** [comment]
The issue probably is `_getEntities`, it takes `hass.states` and thus will return a new array on every state change, this should be made smarter so it only returns a new array if an entity is actually changed

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 1]
```diff
@@ -182,33 +181,47 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
+    if (changedProps.has("hass") || changedProps.has("_config")) {
+      const sourceEntities = this._getSourceEntities(
+        this.hass.states,
+        this._config
+      );
+      if (
+        !this._sourceEntities ||
+        sourceEntities.toString() !== this._sourceEntities.toString()
+      ) {
+        this._sourceEntities = sourceEntities;
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:193]: Should update should not have side effects, you can use `changedProps.get("hass")` to get the previous version of `hass`.

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 2]
```diff
@@ -308,47 +321,47 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
 
-  private _getEntities = memoizeOne(
-    (
-      states: HassEntities,
-      config: MapCardConfig,
-      configEntities?: MapEntityConfig[]
-    ) => {
-      if (!states || !config) {
-        return undefined;
+  private _getSourceEntities = memoizeOne(
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:324]: This should not be memoized, as you are already storing it on the class

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 3]
```diff
@@ -182,33 +181,47 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
-
-    return this._config?.entities
-      ? hasConfigOrEntitiesChanged(this, changedProps)
-      : hasConfigChanged(this, changedProps);
+    return entity_ids.some(
+      (entity_id) =>
+        oldHass.states[entity_id] !== this.hass.states[entity_id] ||
+        oldHass.entities[entity_id]?.display_precision !==
+          this.hass.entities[entity_id]?.display_precision
+    );
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:224]: Why this change?

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 1 above)
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts]: Done. Removed "_sourceEntities" class member.

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 2 above)
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts:325]: Not storing it on the class anymore (removed the class member), so keeping it memoized.

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 3 above)
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts]: Revert these lines. 
(Kept the states comparison separate. One for `this._config.entities` and another one for the source entities.)

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 4]
```diff
@@ -308,47 +320,48 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
-          color: this._getColor(entity),
-        })),
-      ];
+      return entities;
     }
   );
 
+  private _getEntities = memoizeOne(
+    (
+      states: HassEntities,
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:347]: Since `states` is still part of this memoized call, `getEntities` will still return a new array every time a state is changed. I don't think we should solve that is `shouldUpdate`, for example if darkmode changes, this will cause an update and an auto fit, and in the future that could become more problematic.

Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts:347]: Can you elaborate on how you think we should address this issue?
My approach was to refine the old logic [here](https://github.com/home-assistant/frontend/blob/e1dc73e992a620d580cd321f40ffa9ca44b8cc2a/src/panels/lovelace/cards/hui-map-card.ts#L203):
```
if (this._config?.geo_location_sources) {
  if (oldHass.states !== this.hass.states) {
    return true;
  }
}
```
with a comparison of the relevant entities (lines 200-220 in the new code), instead of any entity in the system.

I think that this is a better logic than the old one, regardless of additional improvements we can make in the future. Isn't it?

**Comment by amitfin** [comment]
Below are additional details on the change and its approach.

`shouldUpdate` has 2 places for checking if relevant states have changed:
1. For entities configured in `this._config.entities` it's done by calling `hasConfigOrEntitiesChanged`. The function compares the state of the selected entities. 
2. For entities coming via `this._config.geo_location_sources`: the old implementation checked if there is a state change for any entity in the system. The new implementation is comparing only relevant entities. 

To summarize: the change is to unify `shouldUpdate` logic of state comparison for all entities of `hui-map-card`.

(There was no attempt to make a different use or a change in the `memoizeOne` logic.)

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 4 above)
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts]: `_getEntities` was changed to return the same object when there is no change.

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 5]
```diff
@@ -201,7 +199,22 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
+        oldHass.states,
+        this._config
+      );
+      const newSourceEntities = this._getSourceEntities(
+        this.hass.states,
+        this._config
+      );
+
+      if (
+        oldSourceEntities.toString() !== newSourceEntities.toString() ||
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:212]: Use https://github.com/home-assistant/frontend/blob/dev/src/common/util/deep-equal.ts#L3

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 6]
```diff
@@ -308,46 +321,50 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
+      ...this._getSourceEntities(this.hass.states, this._config).map(
+        (entity) => ({
           entity_id: entity,
           color: this._getColor(entity),
-        })),
-      ];
+        })
+      ),
+    ];
+    if (JSON.stringify(entities) !== JSON.stringify(this._mapEntities)) {
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:363]: use https://github.com/home-assistant/frontend/blob/dev/src/common/util/deep-equal.ts#L3

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 7]
```diff
@@ -70,6 +71,8 @@ class HuiMapCard extends LitElement implements LovelaceCard {
 
   private _configEntities?: MapEntityConfig[];
 
+  private _mapEntities: HaMapEntity[] = [];
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:74]:
```suggestion
  @state() private _mapEntities: HaMapEntity[] = [];
```

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 5 above)
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts]: Done

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 8]
```diff
@@ -156,14 +159,10 @@ class HuiMapCard extends LitElement implements LovelaceCard {
         <div id="root">
           <ha-map
             .hass=${this.hass}
-            .entities=${this._getEntities(
-              this.hass.states,
-              this._config,
-              this._configEntities
-            )}
+            .entities=${this._getMapEntities()}
```
[the remark this answers is not in the document]
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts]: Done

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 7 above)
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts]: Done

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 6 above)
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts]: Done

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 9]
```diff
@@ -308,46 +321,50 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
-      return [
-        ...(configEntities || []).map((entityConf) => ({
-          entity_id: entityConf.entity,
-          color: this._getColor(entityConf.entity),
-          label_mode: entityConf.label_mode,
-          focus: entityConf.focus,
-          name: entityConf.name,
-        })),
-        ...geoEntities.map((entity) => ({
+  private _getMapEntities(): HaMapEntity[] {
```
[the remark this answers is not in the document]
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts]: I would appreciate a guidance here, as I'm not sure where to put this logic.
The call was added to `shouldUpdate()` instead of `render()`, but it's a side-effect. 
The other option is to implement `attributeChangedCallback` and to update `_mapEntities` when `hass` has changed. But there is no use of this callback in the entire codebase as it should be rarely used...

[the remark this answers is not in the document]
Comment (bramkragten) [reply src/panels/lovelace/cards/hui-map-card.ts]: You can use `willUpdate` and check what properties are changed, there you calculate `_mapEntities` and that is used in render.

[the remark this answers is not in the document]
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts]: Done

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 10]
```diff
@@ -202,7 +202,22 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
+        oldHass.states,
+        this._config
+      );
+      const newSourceEntities = this._getSourceEntities(
+        this.hass.states,
+        this._config
+      );
+
+      if (
+        !deepEqual(oldSourceEntities, newSourceEntities) ||
```
[the remark this answers is not in the document]
Comment (bramkragten) [reply src/panels/lovelace/cards/hui-map-card.ts:215]: The render function of this element is really light, so it is not a big deal to call that often, we should just make sure we don't update `_mapEntities` too often.

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 11]
```diff
@@ -309,46 +331,49 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
+      ...this._getSourceEntities(this.hass?.states, this._config).map(
+        (entity) => ({
           entity_id: entity,
           color: this._getColor(entity),
-        })),
-      ];
+        })
+      ),
+    ];
+    if (!deepEqual(entities, this._mapEntities)) {
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:373]: If we move the check from `shouldUpdate` to `willUpdate`, we can remove this.

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts:215]: Can you please elaborate? Should `if (this._config?.geo_location_sources)` block always return true (without checking anything else)?
Otherwise, we need the `deepEqual` to trigger the call to `willUpdate`.

[the remark this answers is not in the document]
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts]: Done

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 11 above)
Comment (amitfin) [reply src/panels/lovelace/cards/hui-map-card.ts]: Done

**Review by bramkragten — APPROVED** [review]

**Comment by karwosts** [comment]
I believe this change broke auto-fit for regular entities, the map no longer follows them.

Previously `ha-map` would run autofit anytime the `entities` property changed, but now it looks like we only update entities on setConfig, and not during each render. 

Should willUpdate not be conditioned on having geo_location_sources?

```
 protected willUpdate(changedProps: PropertyValues): void {
    super.willUpdate(changedProps);
    if (
      changedProps.has("hass") &&
      this._config?.geo_location_sources &&
      !deepEqual(
        this._getSourceEntities(changedProps.get("hass")?.states),
        this._getSourceEntities(this.hass.states)
      )
    ) {
      this._mapEntities = this._getMapEntities();
    }
  }
```
